tests/Medidata.MAuth.Tests/MAuthProtocolSuiteTests.cs

```
@@ -0,0 +1,122 @@
+                Assert.Equal(sig, actualPayload);
+
+
+                // Verify string_to_sign is matched
+                // Act 
+                var result = await new MAuthCoreV2().GetSignature(request, authInfo);
+                var sts = await _protcolTestHelper.ReadStringToSign(caseName);
+
+                // Assert
+                Assert.Equal(sts, Encoding.UTF8.GetString(result));
+            }
+            else
+            {
+                var serverOptions = TestExtensions.ServerOptions;
+                serverOptions.MAuthServerHandler = await new MAuthServerHandler().InitializeAsync();
+
+                var authenticator = new MAuthAuthenticator(
+                    serverOptions, NullLogger<MAuthAuthenticator>.Instance);
+
+                var signedRequest = await new MAuthCoreV2()
+                    .AddAuthenticationInfo(request, new PrivateKeyAuthenticationInfo()
+                    {
+                        ApplicationUuid = signConfig.AppUuid,
+                        PrivateKey = signConfig.PrivateKey,
+                        SignedTime = signConfig.RequestTime.FromUnixTimeSeconds()
+                    });
+
+                // Act
+                var isAuthenticated = await authenticator.AuthenticateRequest(signedRequest);
+
+                // Assert
+                Assert.True(isAuthenticated);
+            }
+        }
+
+        public static IEnumerable<object[]> TestCases
+        {
+            get
+            {
+                // Or this could read from a file. :)
+                var cases = new ProtocolTestSuiteHelper().GetTestCases();
+
+                if (cases is null)
+                    return new List<object[]> { new object[] { string.Empty } };
+
+                var data = new List<object[]>();
+                foreach (string item in cases)
+                {
+                    data.Add(new object[] { item });
+                }
+
+                return data;
+            }
+        }
+    }
```

[nits]
This can be refactored to
```cs
public static IEnumerable<object[]> TestCases =>
    new ProtocolTestSuiteHelper().GetTestCases()
    .Select(tc => new object[] { tc });
```

---

Refactored but kept the cases when `null` -> https://github.com/mdsol/mauth-client-dotnet/pull/64/commits/aa941fa38a149281e66d2b9321d9e2b93b93dcdf#diff-a00da24e99d915ff2dc2ddd89bda839cL102-R115

---

Do we need to check for the existence of the `test cases`?
If there are no `test cases`, what are we even testing?

---

Well, since this test module `mauth-protocol-test-suite` is not being added to this `mauth-client-dotnet` yet unless it is made `public`. Therefore, `mauth-protocol-test-suite` is just cloned locally and when the unit test code finds it runs all 19 test cases. But, it does not find, currently, the unit test compiles for `empty caseName` which just ends the unit test as -> https://github.com/mdsol/mauth-client-dotnet/blob/feature/MCC-575820/tests/Medidata.MAuth.Tests/MAuthProtocolSuiteTests.cs#L29

But, if this part is removed to return empty list of object. And, if any contributor has not cloned the `mauth-protocol-test-suite` then it will fail.
Else have to return `Array.Empty<string>` in else case of `GetTestCases` but it will also yield me :
`System.InvalidOperationException : No data found for Medidata.MAuth.Tests.ProtocolTestSuite.MAuthProtocolSuiteTests.MAuth_Execute_ProtocolTestSuite`

---

So the question is, do we have to run those protocol test suites every time we make changes to mauth-client-dotnet?
If the answer is `yes` then we should fail if we can't read the test cases (due to not cloning the sub-module or other reasons).
Otherwise, maybe it's okay to return empty test data. 

The follow up question is, how do you know your changes are correct if you don't run the protocol test suites?

---

> So the question is, do we have to run those protocol test suites every time we make changes to mauth-client-dotnet?
> If the answer is `yes` then we should fail if we can't read the test cases (due to not cloning the sub-module or other reasons).
> Otherwise, maybe it's okay to return empty test data.

Yes, we have to run protocol test suites if the changes are related to V2 since the protocol test suite is for V2. Since, `mauth-protocol-test-suite` is in the process of being public, we need to skip non "protocol-test-suite" test when just running other unit test. 
Once `mauth-protocol-test-suite` is public, then we do not need to skip the tests any more.
 
> The follow up question is, how do you know your changes are correct if you don't run the protocol test suites?

As mentioned above, we need to run the protocol test suites for the validation of the tests.

Therefore, I will update `CONTRIBUTING.md` how to run non `mauth-protocol-test-suite` test and `mauth-protocol-test-suite` test and remove the empty cases and handling empty cases too.

---

@hkurniawan-mdsol the stance for all the mauth implementations has been to wait until the suite is open sourced and run those tests locally if needs be, see https://mdsol.slack.com/archives/C2MDE04R5/p1594808785184000

---

@prajon84 @jfeltesse-mdsol 

I see the protocol test suites are very important to run against whenever we make changes. If we are shortcutting them due to not able to read the test cases, then changes can lead to a serious bug.

I see that currently they're not open sourced (but will eventually) so we can't add them as a sub-module and if you try to run the whole tests, some will fail. 
Which is fine (and good) since the dev is now aware of the test suites that they need to run against.
If we fail silently, this can be dangerous.